### DISCUSSION TITLE
setAtom uses stale pending on atom unmount

### DISCUSSION THREAD

**Discussion by dmaskasky** [pmndrs/jotai#2812] [Bug report]
### Bug Description

This potentially causes atom listeners and pending unmount functions to not be called.

```ts
const store = createStore()
const a = atom(0)
const b = atom(0, (_, set, update: number) => set(a, update))
b.onMount = (setAtom) => () => setAtom(1)
const aListener = vi.fn()
store.sub(a, aListener)
const unsub = store.sub(b, () => {})
aListener.mockClear()
unsub()
expect(store.get(a)).toBe(1)
expect(aListener).toHaveBeenCalledTimes(1) // received 0
```

### Resolved with

https://github.com/pmndrs/jotai/pull/2811

**Comment by dai-shi** [comment]
~I thought we had a test to cover such case, but obviously there wasn't. Thanks for the catch.~

Wait, it's a `aListener` that should be fired. No, we haven't covered this case before. Yeah, this is reasonable. It's a bug.
